org.metaborg.spoofax.core/src/main/java/org/metaborg/spoofax/core/dynamicclassloading/DynamicClassLoadingService.java

```diff
@@ -0,0 +1,138 @@
+package org.metaborg.spoofax.core.dynamicclassloading;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.Set;
+
+import org.apache.commons.vfs2.FileObject;
+import org.metaborg.core.MetaborgException;
+import org.metaborg.core.language.ILanguageCache;
+import org.metaborg.core.language.ILanguageComponent;
+import org.metaborg.core.language.ILanguageImpl;
+import org.metaborg.core.resource.IResourceService;
+import org.metaborg.util.log.ILogger;
+import org.metaborg.util.log.LoggerUtils;
+
+import com.google.common.collect.Maps;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+
+public class DynamicClassLoadingService implements IDynamicClassLoadingService, ILanguageCache {
+    private static final ILogger logger = LoggerUtils.logger(DynamicClassLoadingService.class);
+
+    private final IResourceService resourceService;
+    private final Set<ClassLoader> additionalClassLoaders;
+
+    private final Map<ILanguageComponent, ClassLoader> classLoaderCache = Maps.newHashMap();
+    @SuppressWarnings("rawtypes")
+    private final Map<ILanguageComponent, Map<Class, ServiceLoader>> serviceLoaderCache = Maps.newHashMap();
+    private final Injector injector;
+
+
+    @Inject public DynamicClassLoadingService(IResourceService resourceService,
+        Set<ClassLoader> additionalClassLoaders, Injector injector) {
+        this.resourceService = resourceService;
+        this.additionalClassLoaders = additionalClassLoaders;
+        this.injector = injector;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> T loadClass(ILanguageComponent component, String className, Class<T> expectedType) throws MetaborgException {
+        final ClassLoader classLoader = classLoader(component);
+        Class<?> theClass;
+        try {
+            logger.trace("Loading outliner class");
+            theClass = classLoader.loadClass(className);
+        } catch (ClassNotFoundException e) {
+            throw new MetaborgException("Given class was not found: " + className, e);
+        }
+        T instance;
+        try {
+            logger.trace("Instantiating outliner class");
+            instance = (T) theClass.newInstance();
+            injector.injectMembers(instance);
+        } catch (InstantiationException e) {
```

Comment:
I'm not a big fan of wrapping all exceptions in a generic MetaborgException.

Reply:
Would you prefer that the Exceptions are just passed up themselves? What do you think of the Exception style @Gohla?

Reply:
There are three concerns here: 1) whether the user of the API needs to handle this exception at all, 2) whether the user of the API needs to differentiate between different kinds of failures, and 3) presenting a nice error message to the user (or developer through logging). 

For 1, you need to think about whether the user of the API can actually recover from these failures. In this case, I don't think the user can, because they all seem to be class loading exceptions (unless I am missing something), so you should consider wrapping them in a runtime exception instead. The user can of course still handle the exception if they want to, by catching the runtime exception. In this case, make sure to document the runtime exception on this method or this method in the interface (if there is one).

For 2, you need to think about whether the user of the API needs to be able to differentiate between different kinds of failures. In this case, since the user of the API probably cannot recover anyway, so differentiating between different exceptions makes no sense.

For 3, in the end, these exceptions should be wrapped with a descriptive error message when presented to the user (or the developer through logging). This can either be a responsibility of this class (i.e., you wrap these exceptions in one with a nicer message like you are doing now), or a responsibility of the user of the API (i.e., the one who calls this function must catch the different exceptions and produce a nice message). In almost all cases, I think the user of the API will just fail/rethrow when a failure occurs, so it is best to wrap them here.

In the end, I would do what you are doing now, but might consider using a runtime exception instead. I would also make the message include a little bit more context: `"Cannot load class '" + className + "`', it was not found", etc.